### DISCUSSION TITLE
could not access s3 without credential secret

### DISCUSSION THREAD

**Discussion by STYNET** [longhorn/longhorn#8474] [General]
Hello,

I have deployed Longhorn v1.6.1 with Helm :
```
$ helm ls
NAME    	NAMESPACE      	REVISION	UPDATED                                 	STATUS  	CHART         	APP VERSION
longhorn	longhorn-system	2       	2024-04-30 09:32:16.981019216 +0200 CEST	deployed	longhorn-1.6.1	v1.6.1 
```

The "backup-target" and  "backup-target-credential-secret" are configured :
```
$ k get settings.longhorn.io backup-target backup-target-credential-secret
NAME                              VALUE                           AGE
backup-target                     s3://longhorn-backup-k8s@sbg/   18h
backup-target-credential-secret   credentials-ovh-s3              18h
```

and the secret "credentials-ovh-s3" for access to my bucket S3 :
```
$ k -n longhorn-system describe secrets  credentials-ovh-s3 
Name:         credentials-ovh-s3
Namespace:    longhorn-system
Labels:       <none>
Annotations:  longhorn.io/backup-target: s3://longhorn-backup-k8s@sbg/

Type:  Opaque

Data
====
AWS_ACCESS_KEY_ID:      32 bytes
AWS_ENDPOINTS:          28 bytes
AWS_SECRET_ACCESS_KEY:  32 bytes
VIRTUAL_HOSTED_STYLE:   4 bytes
```

But, when i click menu "Backup" in the Longhorn UI, i have this message :/
`failed to get default backup target client on longhornio/longhorn-engine:v1.6.1: could not access s3 without credential secret`

This is a bug or i missing anything ?

**Comment by derekbit** [comment]
Can you show the output of `kubectl -n longhorn-system get backuptarget -o yaml`?

**Reply by derekbit** [reply]
BTW, can you check if there are error messages related to the backup target in longhorn-manager pods?

**Reply by STYNET** [reply]
```
$ kubectl -n longhorn-system get backuptarget -o yaml
apiVersion: v1
items:
- apiVersion: longhorn.io/v1beta2
  kind: BackupTarget
  metadata:
    creationTimestamp: "2024-04-29T15:05:54Z"
    generation: 257
    name: default
    namespace: longhorn-system
    resourceVersion: "43383449845"
    uid: 834604de-3306-4c3d-8781-07aad302ba87
  spec:
    backupTargetURL: s3://longhorn-backup-k8s@sbg/
    credentialSecret: credentials-ovh-s3
    pollInterval: 5m0s
    syncRequestedAt: "2024-04-30T13:47:26Z"
  status:
    available: false
    conditions:
    - lastProbeTime: ""
      lastTransitionTime: "2024-04-29T15:06:05Z"
      message: 'failed to get default backup target client on longhornio/longhorn-engine:v1.6.1:
        could not access s3 without credential secret'
      reason: Unavailable
      status: "True"
      type: Unavailable
    lastSyncedAt: null
    ownerID: pool04-node-67c411
kind: List
metadata:
  resourceVersion: ""
```

**Reply by derekbit** [reply]
Weird. longhorn-manager backup target controller thought `backupTarget.Spec.CredentialSecret` is empty ([codes](https://github.com/longhorn/longhorn-manager/blob/master/controller/backup_target_controller.go#L270-L273)).
Can you remove the backup target (s3://longhorn-backup-k8s@sbg/) from the setting `backup-target` and add it back again?

**Reply by derekbit** [reply]
BTW, can you send a support bundle to "longhorn-support-bundle@suse.com"?

**Reply by mantissahz** [reply]
Hi @STYNET,
Could you provide the SB again? I did not see it in the mail you sent.

**Reply by mantissahz** [reply]
It should be related to https://github.com/longhorn/longhorn/issues/8210

**Comment by STYNET** [comment]
logs of longhorn-manager :
```
time="2024-04-30T13:53:43Z" level=warning msg="websocket: close 1000 (normal)" func=api.NewRouter.NewStreamHandlerFunc.func123.1 file="stream.go:52" id=5379273923632466112 type=recurringjobs
91.171.87.244 - - [30/Apr/2024:13:53:40 +0000] "GET /v1/ws/1s/recurringjobs HTTP/1.1" 200 0 "" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:125.0) Gecko/20100101 Firefox/125.0"
91.171.87.244 - - [30/Apr/2024:13:53:43 +0000] "GET /v1/backuptargets HTTP/1.1" 200 518 "https://longhorn.xxxxx.com/" "Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:125.0) Gecko/20100101 Firefox/125.0"
```

**Comment by derekbit** [comment]
cc @mantissahz

**Comment by derekbit** [comment]
@roger-ryao @chriscchien @yangchiu  Did you run into the similar issue before?

**Comment by STYNET** [comment]
I ran several tests yesterday and it seems I've identified bugs in saving backup storage settings.

**Bug 1**: Settings in UI are not always identical to controller values. 
How to reproduce this problem:
- Deploy Longhorn with Helm values (BackupTarget & CredentialSecret")
- View values in UI : 
   UI->ok & Controller->ok 
- From UI, change values (and save): 
   UI->ok & Controller->ok
- From UI, empty values and save (to delete backup storage, for example): 
   UI->  NOK ! :-1: (even after browser refresh) & Controller-> Ok

**Reply by mantissahz** [reply]
@STYNET, 
Can you elaborate on `UI-> NOK`? Does it show the wrong information or show the error logs? 
cc @a110605

**Reply by STYNET** [reply]
UI-> NOK : Settings in UI are not identical to controller values.
- UI : Values not empty
- Controller : Values Empty

**Comment by STYNET** [comment]
**Bug 2**: Messages between UI and controllers are inconsistent

I performed the following tests by modifying the values from the UI:

--- TEST 1 ---
BackupTarget.spec.backupTargetURL: WRONG-URL-FOR-TEST
BackupTarget.spec.credentialSecret: WRONG-SECRET

UI say :
backup target URL is empty

LOGS MANAGER say:
driver  is not supported


--- TEST 2 ---
BackupTarget.spec.backupTargetURL: s3://WRONG-URL-FOR-TEST
BackupTarget.spec.credentialSecret: WRONG-SECRET

UI say :
'failed to get default backup target client on longhornio/longhorn-engine:v1.6.1: secret "WRONG-SECRET" not found'

LOGS MANAGER say:
'failed to get default backup target client on longhornio/longhorn-engine:v1.6.1: secret "WRONG-SECRET" not found'


--- TEST 3 ---
BackupTarget.spec.backupTargetURL: https://s3.sbg.cloud.ovh.net
BackupTarget.spec.credentialSecret: credentials-ovh-s3

UI say :
'failed to get default backup target client on longhornio/longhorn-engine:v1.6.1: secret "WRONG-SECRET" not found'

LOGS MANAGER say:
"Dropping Longhorn backup target out of the queue"
'Failed to run list system backup command'
...

> **Note for test3** : I used my real bucket access values. I've attached the manager logs.
[test3-logs-manager.txt](https://github.com/longhorn/longhorn/files/15185562/test3-logs-manager.txt)

**Comment by derekbit** [comment]
@STYNET Thanks for your findings. 
@mantissahz Can you help create corresponding tickets after clarifying the issues with @STYNET? Thank you.

**Reply by mantissahz** [reply]
Sure, I will try to reproduce them, and find out what happened. And then create the ticket if necessary.

**Comment by mantissahz** [comment]
Hi @STYNET,

Does `could not access s3 without credential secret` still happen?
If it still happened, could you send the support bundle for investigation?

I still can not reproduce `could not access s3 without credential secret` and `Bug 1`.

For `Bug 2`, yeah, we need to have more precise error/warning logs.

**Reply by STYNET** [reply]
Hi Mantissahz,

Thank you for your answers.

Unfortunately, I still have the credential secret problem :-(
I've just emailed the bundle support.

For bug1, here's how to reproduce it (I started from scratch).

**Create secret**
```
export AWS_ACCESS_KEY_ID='xxxxxxxxxx'
export AWS_SECRET_ACCESS_KEY='yyyyyyyyyy'
```

**Create secret for my credentials**
```
kubectl -n longhorn-system create secret generic credentials-ovh-s3 \
  --from-literal=AWS_ACCESS_KEY_ID=${AWS_ACCESS_KEY_ID} \
  --from-literal=AWS_SECRET_ACCESS_KEY=${AWS_SECRET_ACCESS_KEY} \
  --from-literal=AWS_ENDPOINTS=https://s3.sbg.cloud.ovh.net\
  --from-literal=VIRTUAL_HOSTED_STYLE=true\
  -n longhorn-system \
&& unset AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY
```

**Install Longhorn with Helm**
```
helm install longhorn longhorn/longhorn \
  --namespace longhorn-system  \
  --create-namespace \
  --version 1.6.1 \
  --set defaultSettings.backupTarget=s3://longhorn-backup-k8s@sbg/ \
  --set defaultSettings.backupTargetCredentialSecret=credentials-ovh-s3
```

**Other**
#Ingress and basic-auth for UI

**Recreate bug1**

1. Connect on UI app and go to "Setting", "General"
The values is right :
Backup Target : s3://longhorn-backup-k8s@sbg/
Backup Target Credential Secret : credentials-ovh-s3

2. Remove this two values (empty fields) and Save
The values is right :
Backup Target : 
Backup Target Credential Secret : 

3. Refresh page on your browser
The values don't save
Backup Target : s3://longhorn-backup-k8s@sbg/
Backup Target Credential Secret : credentials-ovh-s3
